### ISSUE TITLE
Post edit script takes way much longer to generate for Gutenberg editor

### ISSUE THREAD

**Issue by gziolo** [WordPress/gutenberg#3167]
## Issue Overview

I encountered this issue for the first time when opening the post that has **75 revisions**. It took almost **30 seconds** to generate HTML document with PHP when trying to edit the post using **Gutenberg**. It was working normally for the **Classic** editor - **1.42 s**. When I copied the content of this post to the newly created post with Gutenberg it got back to normal - about the same as with classic. As soon as I added over 20 revisions, processing time has increased to 6 seconds for Gutenberg.

## Steps to Reproduce (for bugs)

1. Create a new post with Gutenberg.
2. Paste the content of the post: https://pastebin.com/R8Zue5kQ.
3. Save draft.
4. Open with the Classic editor and check how long does it take to generate PHP document.
5. Repeat (4) for Gutenberg editor.
6. Add as many revisions as possible, 20 should be enough to notice the difference.
7. Repeat steps (4) and (5) and compare times.

## Expected Behavior

It always takes about the same to generate HTML document with PHP when selecting Classic or Gutenberg editor. The number of revisions shouldn't have any impact.

## Current Behavior

It takes way much longer to generate HTML document with PHP when selecting Gutenberg editor than the Classic editor. The number of revisions seems to have a big impact.

##### Gutenberg

<img width="809" alt="gutenberg-editor" src="https://user-images.githubusercontent.com/699132/32050368-d7018120-ba50-11e7-8b6c-be15d9879dd2.png">

##### Classic

<img width="809" alt="classic-editor" src="https://user-images.githubusercontent.com/699132/32050302-96e79cb4-ba50-11e7-89f1-d87ac9d050a4.png">

## Related Issues and/or PRs

#2822

**Comment by gziolo** [comment]
From @pento:

> The debug bar https://wordpress.org/plugins/debug-bar/ with the slow actions add-on https://wordpress.org/plugins/debug-bar-slow-actions/ would be where I’d start.

From @aduth:

> Would be curious if the REST API endpoint takes as long for the same post with many revisions. Since what’s probably taking time server-side is pre-executing these endpoints, might be symptomatic of a larger problem.

**Comment by mikejhale** [comment]
I'm seeing the same thing (30+ seconds  TTFB for post.php) and am also getting maximum execution time exceeded exceptions about half the time on this line: 

https://github.com/WordPress/gutenberg/blob/master/lib/parser.php#L1549

**Comment by gziolo** [comment]
@dmsnell could you check the parser part that seems to cause troubles here?

**Comment by dmsnell** [comment]
Not sure what I can do to check it - the if WordPress is running every revision through the parser that sounds like another problem. The parser should have a linear correlation with post size, but I don't see how revisions should matter here.

We can insert a test timing on a local dev to see for sure.

**Comment by dmsnell** [comment]
Just for funsies I ran that file through the parser in repeated succession and on my Late 2013 rMBP i7 2.6 GHz on battery it was taking an average of around 54ms for each iteration. Granted, that's without WordPress loaded, so maybe there's some kind of memory cap involved here or some memory leak when we pull all of WordPress and Gutenberg together.

The script I ran is available in a gist; I ran it with 1000 iterations, then 2000 iterations, but not 10,000 iterations as I was on battery.
https://gist.github.com/dmsnell/0849dec331d000e25e649fa351ba2db6
